### PR TITLE
Added AIECreateLogicalLocks

### PR DESCRIPTION
**Pull request by iavssw**
This pass aims to assign lockIDs to AIE.lock operations. The lockID is numbered from the most recent AIE.lock within the same tile. If the lockID exceeds 15 then the pass generates an error and terminates. AIE.lock operations for different tiles are numbered independently. If there exists an existing lockID this pass overwrites the existing lockID generating a warning.

### CODE REVIEW

File: include/aie/AIE.td [hunk 1]
```diff
@@ -938,11 +938,19 @@ def AIE_LockOp: AIE_Op<"lock", []>, Results<(outs Index)> {
   }];
   let arguments = (
     ins Index:$tile,
-        Confined<I32Attr, [IntMinValue<0>, IntMaxValue<15>]>:$lockID
+        OptionalAttr<Confined<I32Attr, [IntMinValue<0>, IntMaxValue<15>]>>:$lockID
+        //Confined<I32Attr, [IntMinValue<0>, IntMaxValue<15>]>:$lockID
```
Comment (stephenneuendorffer) [inline include/aie/AIE.td:942]: Remove comment

File: include/aie/AIE.td [hunk 2]
```diff
@@ -938,11 +938,19 @@ def AIE_LockOp: AIE_Op<"lock", []>, Results<(outs Index)> {
...
   );
-  let assemblyFormat = [{ `(` $tile `,` $lockID `)` attr-dict }];
+  let assemblyFormat = [{ `(` $tile (`,` $lockID^ )? `)` attr-dict }];
+  //let assemblyFormat = [{ `(` $tile `,` $lockID `)` attr-dict }];
```
Comment (stephenneuendorffer) [inline include/aie/AIE.td:945]: comment

File: include/aie/AIEPasses.td [hunk 3]
```diff
@@ -99,6 +99,16 @@ def AIECreateLocks : Pass<"aie-create-locks", "ModuleOp"> {
 
   let constructor = "xilinx::AIE::createAIECreateLocksPass()";
 }
+
+def AIECreateLogicalLocks : Pass<"aie-create-logical-locks", "ModuleOp"> {
+  let summary = "Generates the absolute lock location numbers.";
+  let description = [{
+    Generates the absolute lock location numbers.
```
Comment (stephenneuendorffer) [inline include/aie/AIEPasses.td:106]: Can you add an example of what the IR looks like before and after?

File: include/aie/AIE.td [hunk 4]
```diff
@@ -938,11 +938,19 @@ def AIE_LockOp: AIE_Op<"lock", []>, Results<(outs Index)> {
...
   let extraClassDeclaration = [{
-    int getLockID() { return lockID(); }
+    int getLockID() { 
+      if (lockID().hasValue()) {
+        return lockID().getValue(); 
+      } else {
+        llvm::errs() << "Lock has no ID value\n";
```
Comment (stephenneuendorffer) [inline include/aie/AIE.td:951]: This should give you a warning about not returning a value.  Probably this should assert, at least?

File: include/aie/AIEPasses.td [hunk 5]
```diff
@@ -99,6 +99,16 @@ def AIECreateLocks : Pass<"aie-create-locks", "ModuleOp"> {
 
   let constructor = "xilinx::AIE::createAIECreateLocksPass()";
 }
+
+def AIECreateLogicalLocks : Pass<"aie-create-logical-locks", "ModuleOp"> {
```
Comment (stephenneuendorffer) [inline include/aie/AIEPasses.td:103]: How about "aie-assign-locks" or "aie-assign-lock-ids"?

File: lib/AIECreateLogicalLocks.cpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
+//===- AIEAssignBuffers.cpp -------------------------------------*- C++ -*-===//
```
Comment (stephenneuendorffer) [inline lib/AIECreateLogicalLocks.cpp:1]: wrong name.

File: lib/AIECreateLogicalLocks.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2019 Xilinx Inc.
```
Comment (stephenneuendorffer) [inline lib/AIECreateLogicalLocks.cpp:7]: copyright 2022?

File: lib/AIECreateLogicalLocks.cpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+// This pass aims to assign lockIDs to AIE.lock operations. The lockID is
+// numbered from the most recent AIE.lock within the same tile. If the lockID
+// exceeds 15 then the pass generates an error and terminates. AIE.lock
+// operations for different tiles are numbered independently. If there exists an
+// existing lockID this pass overwrites the existing lockID generating a
```
Comment (stephenneuendorffer) [inline lib/AIECreateLogicalLocks.cpp:15]: I think if there are existing lock IDs, this pass should be idempotent and not change anything.  (i.e., it should only assign lock ids to locks without an ID.)

File: lib/AIECreateLogicalLocks.cpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    registry.insert<xilinx::AIE::AIEDialect>();
+  }
+  void runOnOperation() override {
+
+    ModuleOp m = getOperation();
+    OpBuilder rewriter = OpBuilder::atBlockEnd(m.getBody());
+
+    // loop through locks
+    // store lockID count in map with operation as key
+    std::map<Operation *, int> unique_tiles;
```
Comment (stephenneuendorffer) [inline lib/AIECreateLogicalLocks.cpp:44]: maybe "tileToLastID"?  This gives a little bit more information about what the Operation * is pointing to.

File: test/create-logical-locks/logicallocks.mlir [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+// CHECK:    %4 = AIE.tile(4, 3)
+// CHECK:    %5 = AIE.lock(%0, 0)
+// CHECK:    %6 = AIE.lock(%3, 0)
+// CHECK:    %7 = AIE.lock(%3, 1)
+// CHECK:    %8 = AIE.lock(%4, 0)
+// CHECK:    %9 = AIE.core(%0) {
+// CHECK:      AIE.useLock(%5, Acquire, 0)
+// CHECK:      AIE.useLock(%5, Release, 1)
+// CHECK:      AIE.end
+// CHECK:    }
```
Comment (stephenneuendorffer) [inline test/create-logical-locks/logicallocks.mlir:27]: For the purposes of this test, the useLock operations are irrelevent, I think?  If so, it's usually good practice to reduce the size of the test to the absolute minimum.  Also, there are some behaviors you haven't tested (like what happens when there are already lock ids).

**Review by stephenneuendorffer — CHANGES_REQUESTED** [review]

**Comment by iavssw** [comment]
An implementation of a two pass assignLockIDs pass which is idempotent to existing lockIDs.

**Review by stephenneuendorffer — APPROVED** [review]
Looks reasonable.

**Comment by stephenneuendorffer** [comment]
@iavssw Can you rebase your patch so it can be merged?

**Comment by iavssw** [comment]
@stephenneuendorffer Rebased

**Comment by iavssw** [comment]
Rebased

**Pull request merged**